### DISCUSSION TITLE
How do I set an environment variable for a Lua filter's `os.execute`?

### DISCUSSION THREAD

**Discussion by adamiturabi** [quarto-dev/quarto-cli#11569] [Q&A] [answered]
### Description

I am using this Lua filter for rendering TikZ diagrams: https://github.com/danmackinlay/quarto_tikz/blob/main/_extensions/tikz/tikz.lua

On [this line](https://github.com/danmackinlay/quarto_tikz/blob/6cd699abf22c8b8fd34605dc071d5fcde343915c/_extensions/tikz/tikz.lua#L122), it calls `os.execute()` to run a shell command:

```lua
local _, _, dvisvgmExitCode = os.execute("dvisvgm --font-format=woff " .. dviFile .. " -n -o " .. svgFile)
```

My issue is that `dvisvgm` needs the environment variable `GSLIB` set to the location of a library on my system (`LIBGS=/usr/local/Cellar/ghostscript/10.04.0/lib/libgs.dylib`).

How may I have this environment variable set and available to `os.execute()`?

Lua does not seem to have a `os.setenv()` function.

I've tried the following unsuccessfully:

```lua
local _, _, dvisvgmExitCode = os.execute("LIBGS=/usr/local/Cellar/ghostscript/10.04.0/lib/libgs.dylib dvisvgm --font-format=woff " .. dviFile .. " -n -o " .. svgFile)
```

```lua
local _, _, dvisvgmExitCode = os.execute("export LIBGS=/usr/local/Cellar/ghostscript/10.04.0/lib/libgs.dylib; dvisvgm --font-format=woff " .. dviFile .. " -n -o " .. svgFile)
```

```lua
local _, _, dvisvgmExitCode = os.execute("export LIBGS=/usr/local/Cellar/ghostscript/10.04.0/lib/libgs.dylib && dvisvgm --font-format=woff " .. dviFile .. " -n -o " .. svgFile)
```

EDIT:

I found a workaround to use the `--libgs=` option for `dvisvgm` thus:

```lua
local _, _, dvisvgmExitCode = os.execute("dvisvgm --libgs=/usr/local/Cellar/ghostscript/10.04.0/lib/libgs.dylib --font-format=woff " .. dviFile .. " -n -o " .. svgFile)
```

So I'm unblocked. Not sure if the original question still merits an answer.

**Comment by tarleb** [comment, marked as answer]
Pandoc offers the function [`pandoc.system.with_environment`](https://pandoc.org/lua-filters#pandoc.system.with_environment). The function doesn't seem to work reliably on Windows, but it might be good enough for many use-cases.

Usage:

``` lua
pandoc.system.with_environment(
  { LIBGS = "...." },
  function ()
    os.execute('...')
  end
)
```

**Reply by adamiturabi** [reply]
Thank you. This worked.

But there was an issue with the filter where it was picking up an (incorrectly generated) cached version of the svg file instead of generating a new one. When I remove the cache, I notice that this solution also works:

```lua
local _, _, dvisvgmExitCode = os.execute("LIBGS=/usr/local/Cellar/ghostscript/10.04.0/lib/libgs.dylib dvisvgm --font-format=woff " .. dviFile .. " -n -o " .. svgFile)
```
